barretenberg/cpp/src/barretenberg/dsl/acir_format/honk_recursion_constraint.cpp

```diff
@@ -171,76 +43,59 @@ HonkRecursionConstraintOutput<typename Flavor::CircuitBuilder> create_honk_recur
                                   stdlib::recursion::honk::RollupIO,
                                   stdlib::recursion::honk::DefaultIO<Builder>>;
 
-    BB_ASSERT(input.proof_type == HONK || input.proof_type == HONK_ZK || HasIPAAccumulator<Flavor>);
-    BB_ASSERT_EQ(input.proof_type == ROLLUP_HONK || input.proof_type == ROOT_ROLLUP_HONK, HasIPAAccumulator<Flavor>);
-
-    // Construct an in-circuit representation of the verification key.
-    // For now, the v-key is a circuit constant and is fixed for the circuit.
-    // (We may need a separate recursion opcode for this to vary, or add more config witnesses to this opcode)
-    std::vector<field_ct<Builder>> vk_fields = fields_from_witnesses(builder, input.key);
-
-    // Create circuit type for vkey hash.
-    auto vk_hash = field_ct<Builder>::from_witness_index(&builder, input.key_hash);
+    BB_ASSERT(input.proof_type == HONK || input.proof_type == HONK_ZK || input.proof_type == ROLLUP_HONK ||
+                  input.proof_type == ROOT_ROLLUP_HONK,
+              "create_honk_recursion_constraints: Only HONK, HONK_ZK, ROLLUP_HONK, ROOT_ROLLUP_HONK proof types are "
+              "supported.");
+    BB_ASSERT_EQ(input.proof_type == ROLLUP_HONK || input.proof_type == ROOT_ROLLUP_HONK,
+                 HasIPAAccumulator<Flavor>,
+                 "create_honk_recursion_constraints: ROLLUP_HONK and ROOT_ROLLUP_HONK must be recursively verified "
+                 "using a Flavor with IPA accumulator.s");
+
+    // Step 1.
+    // Construct in-circuit representations of the recursion data
+    std::vector<field_ct> vk_fields = fields_from_witnesses(builder, input.key);
+    field_ct vk_hash = field_ct::from_witness_index(&builder, input.key_hash);
+    std::vector<field_ct> proof_fields =
+        fields_from_witnesses(builder, add_public_inputs_to_proof(input.proof, input.public_inputs));
+    bool_ct predicate(to_field_ct(input.predicate, builder)); // Constructor enforces predicate = 0 or 1
+
+    // Step 2. and 3.
+    // Construct an Honk proof and vk with the correct number of public inputs.
+    const auto [honk_proof, honk_vk] = construct_honk_proof_for_simple_circuit<typename Flavor::NativeFlavor>(
```

Comment:
I wonder if we could further improve some of the naming here in line with what you've done further down, e.g.

```suggestion
    const auto [valid_proof, valid_vk] = construct_arbitrary_valid_honk_proof_and_vk<typename Flavor::NativeFlavor>(
```
I think this would be more precise and also make things like `populate_fields(builder, proof_fields, honk_proof);` a bit easier to read, i.e. `populate_fields(builder, proof_fields, valid_proof);`